data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java

```diff
@@ -75,43 +75,62 @@ public DspCatalogApiController(Monitor monitor, ObjectMapper mapper, IdentitySer
+            return errorResponse(Response.Status.BAD_REQUEST, BAD_REQUEST);
         }
+
         var expandedJson = expanded.getContent();
         if (!isOfExpectedType(expandedJson, DSPACE_CATALOG_REQUEST_TYPE)) {
-            throw new InvalidRequestException(format("Request body was not of expected type: %s", DSPACE_CATALOG_REQUEST_TYPE));
+            return errorResponse(Response.Status.BAD_REQUEST, BAD_REQUEST);
         }
 
-        var message = transformerRegistry.transform(expandedJson, CatalogRequestMessage.class)
-                .orElseThrow(failure -> new InvalidRequestException(format("Request body was malformed: %s", failure.getFailureDetail())));
+        var transformResult = transformerRegistry.transform(expandedJson, CatalogRequestMessage.class);
+        if (transformResult.failed()) {
+            return errorResponse(Response.Status.BAD_REQUEST, BAD_REQUEST);
+        }
 
+        var message = transformResult.getContent();
         // set protocol
         message.setProtocol(DATASPACE_PROTOCOL_HTTP);
 
-        var catalog = service.getCatalog(message, claimToken)
-                .orElseThrow(exceptionMapper(Catalog.class));
+        var claimToken = verificationResult.getContent();
 
-        var catalogJson = transformerRegistry.transform(catalog, JsonObject.class)
-                .orElseThrow(failure -> new EdcException(format("Failed to build response: %s", failure.getFailureDetail())));
+        Catalog catalog;
+        try {
+            catalog = service.getCatalog(message, claimToken).getContent();
+        } catch (Exception exception) {
+            var errorCode = UUID.randomUUID();
+            monitor.warning(String.format("DSP negotiation error id %s: %s", errorCode, exception));
+            return errorResponse(Response.Status.INTERNAL_SERVER_ERROR, String.format("Error code %s", errorCode));
+        }
 
-        var compacted = jsonLdService.compact(catalogJson);
-        if (compacted.failed()) {
-            throw new InvalidRequestException(compacted.getFailureDetail());
+        var catalogJson = transformerRegistry.transform(catalog, JsonObject.class);
+        if (catalogJson.failed()) {
+            var errorCode = UUID.randomUUID();
+            monitor.warning(String.format("DSP negotiation error id %s: %s", errorCode, catalogJson.getFailureMessages()));
+            return errorResponse(Response.Status.INTERNAL_SERVER_ERROR, String.format("Error code %s", errorCode));
         }
-        //noinspection unchecked
-        return mapper.convertValue(compacted.getContent(), Map.class);
+
+        return Response.status(Response.Status.OK).type(MediaType.APPLICATION_JSON).entity(catalogJson.getContent()).build();
+    }
+
+    private Response errorResponse(Response.Status code, String message) {
+        return Response.status(code).type(MediaType.APPLICATION_JSON)
+                .entity(DspError.create(DSPACE_CATALOG_ERROR, Optional.empty(), String.valueOf(code.getStatusCode()), message))
```

Comment:
DspError should be a builder. This will remove the need to pass Optional and make the interface fluent for adding problem messages.

Also, even though, in this case, code cannot be null, please don't use String.valueOf() because I came across a number of areas where it was in the codebase, which lead to difficult-to-find error because it converts null to "null."  toString() should be sufficient